plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/BaseImplementAvgBigint.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.jdbc.expression;
+
+import io.trino.matching.Capture;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.plugin.base.expression.AggregateFunctionRule;
+import io.trino.plugin.jdbc.JdbcColumnHandle;
+import io.trino.plugin.jdbc.JdbcExpression;
+import io.trino.plugin.jdbc.JdbcTypeHandle;
+import io.trino.spi.connector.AggregateFunction;
+import io.trino.spi.expression.Variable;
+
+import java.sql.Types;
+import java.util.Optional;
+
+import static com.google.common.base.Verify.verify;
+import static io.trino.matching.Capture.newCapture;
+import static io.trino.plugin.base.expression.AggregateFunctionPatterns.basicAggregation;
+import static io.trino.plugin.base.expression.AggregateFunctionPatterns.expressionType;
+import static io.trino.plugin.base.expression.AggregateFunctionPatterns.functionName;
+import static io.trino.plugin.base.expression.AggregateFunctionPatterns.singleInput;
+import static io.trino.plugin.base.expression.AggregateFunctionPatterns.variable;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.DoubleType.DOUBLE;
+
+/**
+ * Implements {@code avg(x)} for bigint columns while preserving Trino semantics.
+ * Trino semantics say the output should be a double but pushing down the aggregation to some databases
+ * can result in rounding of the output to a bigint.
+ */
+public abstract class BaseImplementAvgBigint
+        implements AggregateFunctionRule<JdbcExpression>
+{
+    private final Capture<Variable> input;
+
+    public BaseImplementAvgBigint()
+    {
+        this.input = newCapture();
+    }
+
+    protected abstract String pushedExpression(String columnName);
```

Comment:
move to the bottom and document

is it actually equivalent to JdbcClient.quote(columnName)?

Reply:
No, the purpose of this method is to do the database-specific implementation of pushing down the `avg()` function over a bigint col. 

Some dbs support `avg(cast(columnName as double))` but some do not [eg: older versions of mysql] so a different approach needs to be taken like `avg(columnName * 1.0)`

Documented.